### PR TITLE
feat: add --tag-prefix, --tag-pattern and --compare-url options

### PR DESCRIPTION
**Pull request by LeMimit**
Hello !

I added the following options :

- `--tag-prefix`: prefix used in version tags, default: "" 

- `--tag-pattern`: override regex pattern for release tags.
  By default, I using semver tag names ([spec](https://semver.org/)). 
For the pattern, I based my code on the regex provided in the spec ([link](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string)).
The regex provided must contains a group named "version" which represent the version.
Exemple: `"(?P<version>[1-9])"`

- `--compare-url`: override url for compares, use `{current}` and `{previous}` for tags. 
The code automatically detects the release title in the changelog based on the `tag prefix` and `tag pattern` and insert the compare url.

I created new tests in order to test this new options !

### CODE REVIEW

File: tests/test_repository.py [hunk 1]
```diff
@@ -37,14 +37,57 @@ def test_index_init():
...
+
+    selected_tags = []
+    for selected_tag_ref_list in selected_tag_refs.values():
+        for selected_tag_ref in selected_tag_ref_list:
+            selected_tags.append(selected_tag_ref.name)
+
+    if len(selected_tags) == 0:
+        assert len(expected_tags) == 0
+    else:
+        assert all([a == b for a, b in zip(selected_tags, expected_tags)])
```
Comment (KeNaCo) [inline tests/test_repository.py]: Why you don't just `assert expected_tags == selected_tags`?

**Comment by KeNaCo** [comment]
I was not sure about diff link, so I try to look at its output and I'm not sure if it works properly..

Screenshot from 2019-10-26 06-38-31

**Comment by LeMimit** [comment]
Hello !

Indeed, the compare url was inverted in my code. My bad !
I fixed the problem.

File: tests/test_repository.py (hunk 1 above)
Comment (LeMimit) [reply tests/test_repository.py]: Yes, you are right. I was going all in with a way more complicated solution.
I fixed it.

**Comment by LeMimit** [comment]
Hello !

Indeed, the compare url was inverted in my code. My bad !
I fixed the problem.

Moreover, I rebased my branch to take last modifications of master branch and I added more use cases to the tests of `compare-url` option.
Thanks to this these new use cases, I found two problems and fixed it !

**Comment by KeNaCo** [comment]
I was thinking from the morning about `--compare-url` feature and have one proposal.

I'm not sure that link from a tag is the obvious way to present this. For example, I will expect this link for [v0.3.0](https://github.com/Michael-F-Bryan/auto-changelog/releases/tag/0.3.0).
My idea is to create separate link under tag:
```
# 0.3.0

_[Source Diff](https://github.com/LeMimit/auto-changelog/compare/0.1.7...0.3.0)_
```
Then we may rename `--compare-url` to `--diff-url`. If not present, we don't include Source diff link. And with parameter will overwrite default generated from remote URL(works same way as `--issue-url`)

What do you think?

**Comment by LeMimit** [comment]
Hello ! 

yes I think this is a good idea. I will look into it today.

**Comment by LeMimit** [comment]
Hello !

I changed how is managed diff url following your suggestion.
The diff url is now contains in the release object. The display is now managed by the template. 
This way the format can be changed by a custom template with the user wants to.
I managed diff url for unreleased changes, in this case the diff url is created with HEAD reference. I had to differentiate the title of a release and its tag in order to do it.

Btw I rebased the branch with the lastest changes in master.

**Comment by LeMimit** [comment]
Hello !

Any news on when you can have a look to merge this PR ? 👍 

I am planning to use it soon.

Thanks a lot !

**Comment by KeNaCo** [comment]
Hi,
I looked at it, 
then have some ideas, 
then think that tests first may be a good idea,
then rewrite tests and introduce new dependency,
then include py3.8,
then discover that new dependency is not on py3.8, yet
then w8,
then Xmas happen and traveling, family, stuff around,
then still not migrated, so w8,
then is today. Dependency still not migrated.

Battleplan:
1. look at your branch and mine
2. separate tests from coverage requirement
3. rebase-retest-merge everything
4. release :confetti_ball: 

No promises... I'll try to finish before the weekend.

**Comment by LeMimit** [comment]
Hello ! 

This is cool to split this PR ! I rebased this PR with the lastest change just in case we want to use it at any point.

@KeNaCo I see the PR for `--tag-prefix` and `--tag-pattern` but not for `--diff-url`. I assume you want to do another PR to isolate the change.
Do you want me to do it ? I can create a new PR with only the change linked to `--diff-url` when #55 and #56 are merged.

**Pull request closed without merging**